java/src/main/java/ai/rapids/cudf/DType.java

```diff
@@ -17,76 +17,202 @@
+  public static final DType DURATION_NANOSECONDS = new DType(DTypeEnum.DURATION_NANOSECONDS);
+  public static final DType STRING = new DType(DTypeEnum.STRING);
+  public static final DType LIST = new DType(DTypeEnum.LIST);
+  public static final DType STRUCT = new DType(DTypeEnum.STRUCT);
+
+  private static final DType[] staticDTypes = new DType[]{
+      EMPTY,
+      INT8,
+      INT16,
+      INT32,
+      INT64,
+      UINT8,
+      UINT16,
+      UINT32,
+      UINT64,
+      FLOAT32,
+      FLOAT64,
+      BOOL8,
+      TIMESTAMP_DAYS,
+      TIMESTAMP_SECONDS,
+      TIMESTAMP_MILLISECONDS,
+      TIMESTAMP_MICROSECONDS,
+      TIMESTAMP_NANOSECONDS,
+      DURATION_DAYS,
+      DURATION_SECONDS,
+      DURATION_MILLISECONDS,
+      DURATION_MICROSECONDS,
+      DURATION_NANOSECONDS,
+      null, // DICTIONARY32
+      STRING,
+      LIST,
+      null, // DECIMAL32
+      null, // DECIMAL64
+      STRUCT
+  };
+
   /**
-   * ms since the UNIX epoch
+   * This only works for fixed width types. Variable width types like strings the value is
+   * undefined and should be ignored.
+   *
+   * @return size of type in bytes.
    */
-  TIMESTAMP_MILLISECONDS(8, 14, "timestamp[ms]"),
+  public int getSizeInBytes() { return typeId.sizeInBytes; }
+
   /**
-   * microseconds since the UNIX epoch
+   * Returns
+   * @return
    */
-  TIMESTAMP_MICROSECONDS(8, 15, "timestamp[us]"),
+  public int getNativeId() { return typeId.nativeId; }
+
   /**
-   * ns since the UNIX epoch
+   * Returns scale for Decimal Type
+   * @return scale
    */
-  TIMESTAMP_NANOSECONDS(8, 16, "timestamp[ns]"),
-
-  //We currently don't have mappings for duration type to I/O files, and these
-  //simpleNames might change in future when we do
-  DURATION_DAYS(4, 17, "int32"),
-  DURATION_SECONDS(8, 18, "int64"),
-  DURATION_MILLISECONDS(8, 19, "int64"),
-  DURATION_MICROSECONDS(8, 20, "int64"),
-  DURATION_NANOSECONDS(8, 21, "int64"),
-  //DICTIONARY32(4, 22, "NO IDEA"),
-
-  STRING(0, 23, "str"),
-  LIST(0, 24, "list"),
-  STRUCT(0, 27, "struct");
-
-  private static final DType[] D_TYPES = DType.values();
-  final int sizeInBytes;
-  final int nativeId;
-  final String simpleName;
-
-  DType(int sizeInBytes, int nativeId, String simpleName) {
-    this.sizeInBytes = sizeInBytes;
-    this.nativeId = nativeId;
-    this.simpleName = simpleName;
+  public int getScale() { return scale; }
+
+  public String getSimpleName() { return typeId.simpleName; }
+
+  // TODO: override hashCode
+  @Override
+  public boolean equals(Object o) {
+    DType other = (DType) o;
+    return this.typeId == other.typeId && this.scale == other.scale;
   }
 
-  public boolean isTimestamp() {
-    return TIMESTAMPS.contains(this);
+  public static DType fromNative(int nativeId, int scale) {
+    if (nativeId >=0 && nativeId < staticDTypes.length) {
+      if (staticDTypes[nativeId] != null) {
+        return staticDTypes[nativeId];
```

Comment:
I'm guessing @revans2 would like the original logic preserved where it would scan the types checking for the first one that matched native IDs.  At a minimum it needs to verify that the native ID of the DType enum it loaded matches the one specified and throw if it doesn't (meaning there's a programming error on the order of the table).  I think it's very likely someone will accidentally update one of these tables without the other (e.g.: add a new DType enum value and forget to put a `null` in the corresponding singleton array.  Comments above each of these lists stating it's important to keep the other list in sync would help.